### PR TITLE
[test] add e2e test about egress access from ns on one port

### PR DESCRIPTION
**Pull request by yulng**
testcase: engress access from ns on one named port
networkpolicy: 
the port 80 is allowed to namespace nxY
the port 81 is blocked for all namespaces including nxY
the previous test case has  allow egress access on one named port, but there's no restriction on namespace

### CODE REVIEW

File: test/e2e/network/netpol/network_policy.go [hunk 1]
```diff
@@ -571,6 +571,36 @@ var _ = common.SIGDescribe("Netpol", func() {
 			ValidateOrFail(k8s, &TestCase{ToPort: 81, Protocol: v1.ProtocolTCP, Reachability: reachabilityPort81})
 		})
 
+		ginkgo.It("should allow engress access from namespace on one named port [Feature:NetworkPolicy]", func(ctx context.Context) {
```
Comment (aojea) [inline test/e2e/network/netpol/network_policy.go:574]:
```suggestion
		ginkgo.It("should allow egress access from namespace on one named port [Feature:NetworkPolicy]", func(ctx context.Context) {
```

Comment (yulng) [reply test/e2e/network/netpol/network_policy.go:574]: changed,thanks

**Comment by jayunit100** [comment]
Thanks for this will review it today
Cc netpol testing crew 
 @mattfenwick

File: test/e2e/network/netpol/network_policy.go [hunk 2]
```diff
@@ -571,6 +571,36 @@ var _ = common.SIGDescribe("Netpol", func() {
...
+			nsX, nsY, nsZ := getK8sNamespaces(k8s)
+			allowedLabels := &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					namespaceLabelKey: nsY,
+				},
+			}
+			engressRule := networkingv1.NetworkPolicyEgressRule{}
+			engressRule.Ports = append(engressRule.Ports, networkingv1.NetworkPolicyPort{Port: &intstr.IntOrString{Type: intstr.String, StrVal: "serve-80-tcp"}})
+			engressRule.To = append(engressRule.To, networkingv1.NetworkPolicyPeer{NamespaceSelector: allowedLabels})
+			policy := GenNetworkPolicyWithNameAndPodMatchLabel("allow-client-a-via-ns-selector-80", map[string]string{"pod": "a"}, SetSpecEgressRules(engressRule))
```
Comment (jayunit100) [inline test/e2e/network/netpol/network_policy.go:587]: s/engress/egrees

File: test/e2e/network/netpol/network_policy.go [hunk 3]
```diff
@@ -571,6 +571,36 @@ var _ = common.SIGDescribe("Netpol", func() {
...
+			CreatePolicy(ctx, k8s, policy, nsX)
+
+			reachabilityPort80 := NewReachability(k8s.AllPodStrings(), true)
+			// disallow all traffic from the x or z namespaces
+			reachabilityPort80.ExpectPeer(&Peer{Namespace: nsX}, &Peer{Namespace: nsX, Pod: "a"}, false)
+			reachabilityPort80.ExpectPeer(&Peer{Namespace: nsZ}, &Peer{Namespace: nsX, Pod: "a"}, false)
+			ginkgo.By("Verify that port 80 is allowed for namespace nxY")
+			ValidateOrFail(k8s, &TestCase{ToPort: 80, Protocol: v1.ProtocolTCP, Reachability: reachabilityPort80})
+
+			// meanwhile no traffic over 81 should work, since our engress policy is on 80
```
Comment (jayunit100) [inline test/e2e/network/netpol/network_policy.go:597]:
```suggestion
			// traffic to X over port 81 should fail
```

File: test/e2e/network/netpol/network_policy.go [hunk 4]
```diff
@@ -571,6 +571,36 @@ var _ = common.SIGDescribe("Netpol", func() {
...
+			protocols := []v1.Protocol{protocolTCP}
+			ports := []int32{80, 81}
+			k8s = initializeResources(ctx, f, protocols, ports)
+			nsX, nsY, nsZ := getK8sNamespaces(k8s)
+			allowedLabels := &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					namespaceLabelKey: nsY,
+				},
+			}
+			engressRule := networkingv1.NetworkPolicyEgressRule{}
```
Comment (jayunit100) [inline test/e2e/network/netpol/network_policy.go:584]:
```suggestion
			egressRule := networkingv1.NetworkPolicyEgressRule{}
```

Comment (yulng) [reply test/e2e/network/netpol/network_policy.go]: All the above have been corrected

**Comment by jayunit100** [comment]
Looks like netpol e2es failed so maybe there's a logical bug or else you found a bug in a  cni provider... likely the former ?

**Comment by yulng** [comment]
> Looks like netpol e2es failed so maybe there's a logical bug or else you found a bug in a cni provider... likely the former ?
Could  the developer of this module help take a look?

**Comment by mattfenwick** [comment]
Took a look at the error messages , they didn't seem to be related to this pr so let's see what happens when we retest.

**Comment by aojea** [comment]
it seems this exact test is failing consistently
```
should allow egress access from namespace on one named port while blocking other namespaces and port
```

**Comment by jayunit100** [comment]
@yulcan can you make sure the test passes locally?  we'll do the next round of code review . These ci jobs use calico....you can try different cni providers using something like https://antrea.io/docs/main/docs/kind/

**Comment by yulng** [comment]
> @yulcan can you make sure the test passes locally? Then we'll do the next round of code review . These ci jobs use calico....you can try different cni providers using something like https://antrea.io/docs/main/docs/kind/

I don't have a local environment, so I submitted it and ran on github
Is there a script in the k8s project that can create a kind cluster and run e2e test code?

**Comment by jayunit100** [comment]
Here's what I use https://github.com/jayunit100/k8sprototypes/tree/master/kind

It basically installs kind locally with a cni provider like calico or antrea .  You can try both to confirm wether the errors you are seeing are bc of CNI issues or because of the code .......

**Comment by jayunit100** [comment]
heres the test output i ran it locally: 
```
                                                                                                                                                                                                                                                                                                                                                                                                                                                                          
  Apr  1 13:42:50.069: INFO: expected:                                                                                                                                                                                                        
                                                                                                                                                                                                                                              
  -             netpol-x-3300/a netpol-x-3300/b netpol-x-3300/c netpol-y-8025/a netpol-y-8025/b netpol-y-8025/c netpol-z-1862/a netpol-z-1862/b netpol-z-1862/c                                                                               
  netpol-x-3300/a       X               .               .               .               .               .               .               .               .                                                                                     
  netpol-x-3300/b       X               .               .               .               .               .               .               .               .                                                                                     
  netpol-x-3300/c       X               .               .               .               .               .               .               .               .                                                                                     
  netpol-y-8025/a       .               .               .               .               .               .               .               .               .                                                                                     
  netpol-y-8025/b       .               .               .               .               .               .               .               .               .                                                                                     
  netpol-y-8025/c       .               .               .               .               .               .               .               .               .                                                                                     
  netpol-z-1862/a       X               .               .               .               .               .               .               .               .                                                                                     
  netpol-z-1862/b       X               .               .               .               .               .               .               .               .                                                                                     
  netpol-z-1862/c       X               .               .               .               .               .               .               .               .

                                                           
                                                                                                                                                                                                                                              
  Apr  1 13:42:50.069: INFO: observed:      

  -             netpol-x-3300/a netpol-x-3300/b netpol-x-3300/c netpol-y-8025/a netpol-y-8025/b netpol-y-8025/c netpol-z-1862/a netpol-z-1862/b netpol-z-1862/c
  netpol-x-3300/a       X               X               X               .               .               .               X               X               X
  netpol-x-3300/b       .               .               .               .               .               .               .               .               .
  netpol-x-3300/c       .               .               .               .               .               .               .               .               .
  netpol-y-8025/a       .               .               .               .               .               .               .               .               .
  netpol-y-8025/b       .               .               .               .               .               .               .               .               .
  netpol-y-8025/c       .               .               .               .               .               .               .               .               .
  netpol-z-1862/a       .               .               .               .               .               .               .               .               .
  netpol-z-1862/b       .               .               .               .               .               .               .               .               .
  netpol-z-1862/c       .               .               .               .               .               .               .               .               .
                                                           


  Apr  1 13:42:50.069: INFO: comparison:

  -             netpol-x-3300/a netpol-x-3300/b netpol-x-3300/c netpol-y-8025/a netpol-y-8025/b netpol-y-8025/c netpol-z-1862/a netpol-z-1862/b netpol-z-1862/c
  netpol-x-3300/a       .               X               X               .               .               .               X               X               X
  netpol-x-3300/b       X               .               .               .               .               .               .               .               .
  netpol-x-3300/c       X               .               .               .               .               .               .               .               .
  netpol-y-8025/a       .               .               .               .               .               .               .               .               .
  netpol-y-8025/b       .               .               .               .               .               .               .               .               .
  netpol-y-8025/c       .               .               .               .               .               .               .               .               .
  netpol-z-1862/a       X               .               .               .               .               .               .               .               .
  netpol-z-1862/b       X               .               .               .               .               .               .               .               .
  netpol-z-1862/c       X               .               .               .               .               .               .               .               .

```

File: test/e2e/network/netpol/network_policy.go [hunk 5]
```diff
@@ -571,6 +571,37 @@ var _ = common.SIGDescribe("Netpol", func() {
...
+			egressRule := networkingv1.NetworkPolicyEgressRule{}
+			egressRule.Ports = append(egressRule.Ports, networkingv1.NetworkPolicyPort{Port: &intstr.IntOrString{Type: intstr.String, StrVal: "serve-80-tcp"}})
+			egressRule.To = append(egressRule.To, networkingv1.NetworkPolicyPeer{NamespaceSelector: allowedLabels})
+			policy := GenNetworkPolicyWithNameAndPodMatchLabel("allow-client-a-via-ns-selector-80", map[string]string{"pod": "a"}, SetSpecEgressRules(egressRule))
+			CreatePolicy(ctx, k8s, policy, nsX)
+
+			reachabilityPort80 := NewReachability(k8s.AllPodStrings(), true)
+			// disallow all traffic from the x or z namespaces
+			reachabilityPort80.ExpectPeer(&Peer{Namespace: nsX}, &Peer{Namespace: nsX, Pod: "a"}, false)
+			reachabilityPort80.ExpectPeer(&Peer{Namespace: nsZ}, &Peer{Namespace: nsX, Pod: "a"}, false)
```
Comment (jayunit100) [inline test/e2e/network/netpol/network_policy.go:593]: lets get rid of this : the policy you have isnt blocking ingress from z->x

File: test/e2e/network/netpol/network_policy.go [hunk 6]
```diff
@@ -571,6 +571,37 @@ var _ = common.SIGDescribe("Netpol", func() {
...
+			}
+			egressRule := networkingv1.NetworkPolicyEgressRule{}
+			egressRule.Ports = append(egressRule.Ports, networkingv1.NetworkPolicyPort{Port: &intstr.IntOrString{Type: intstr.String, StrVal: "serve-80-tcp"}})
+			egressRule.To = append(egressRule.To, networkingv1.NetworkPolicyPeer{NamespaceSelector: allowedLabels})
+			policy := GenNetworkPolicyWithNameAndPodMatchLabel("allow-client-a-via-ns-selector-80", map[string]string{"pod": "a"}, SetSpecEgressRules(egressRule))
+			CreatePolicy(ctx, k8s, policy, nsX)
+
+			reachabilityPort80 := NewReachability(k8s.AllPodStrings(), true)
+			// disallow all traffic from the x or z namespaces
+			reachabilityPort80.ExpectPeer(&Peer{Namespace: nsX}, &Peer{Namespace: nsX, Pod: "a"}, false)
```
Comment (jayunit100) [inline test/e2e/network/netpol/network_policy.go:592]: get rid of this one, you arent blocking ingress into x/a, but rather, blocking egress out of a

File: test/e2e/network/netpol/network_policy.go [hunk 7]
```diff
@@ -571,6 +571,37 @@ var _ = common.SIGDescribe("Netpol", func() {
...
+				MatchLabels: map[string]string{
+					namespaceLabelKey: nsY,
+				},
+			}
+			egressRule := networkingv1.NetworkPolicyEgressRule{}
+			egressRule.Ports = append(egressRule.Ports, networkingv1.NetworkPolicyPort{Port: &intstr.IntOrString{Type: intstr.String, StrVal: "serve-80-tcp"}})
+			egressRule.To = append(egressRule.To, networkingv1.NetworkPolicyPeer{NamespaceSelector: allowedLabels})
+			policy := GenNetworkPolicyWithNameAndPodMatchLabel("allow-client-a-via-ns-selector-80", map[string]string{"pod": "a"}, SetSpecEgressRules(egressRule))
+			CreatePolicy(ctx, k8s, policy, nsX)
+
```
Comment (jayunit100) [inline test/e2e/network/netpol/network_policy.go:589]: i think what you wanted here was 

```
                        
                        CreatePolicy(ctx, k8s, policy, nsX)

                        reachabilityPort80 := NewReachability(k8s.AllPodStrings(), true)
                        // disallow all traffic to the X or Z namespace, BUT allow Y
                        reachabilityPort80.ExpectPeer(&Peer{Namespace: nsX, Pod:"a"}, &Peer{Namespace: nsX}, false)
                        reachabilityPort80.ExpectPeer(&Peer{Namespace: nsX, Pod:"a"}, &Peer{Namespace: nsZ}, false)
                        
                        //reachabilityPort80.ExpectPeer(&Peer{Namespace: nsZ}, &Peer{Namespace: nsX, Pod: "a"}, false)

                        ginkgo.By("Verify that port 80 is from ns y, but not from namespaces x and z")
                        // traffic to X over port 80 should succeed
                        ValidateOrFail(k8s, &TestCase{ToPort: 80, Protocol: v1.ProtocolTCP, Reachability: reachabilityPort80})

                        // traffic to X over port 81 should fail
                        ginkgo.By("Verify that port 81 is blocked for all namespaces including nxY")
                        reachabilityPort81 := NewReachability(k8s.AllPodStrings(), true)
                        reachabilityPort81.ExpectAllEgress(NewPodString(nsX, "a"), false)
                        ValidateOrFail(k8s, &TestCase{ToPort: 81, Protocol: v1.ProtocolTCP, Reachability: reachabilityPort81})

```

**Comment by jayunit100** [comment]
ok, i have a fix you can use 

```
                        
                        CreatePolicy(ctx, k8s, policy, nsX)

                        reachabilityPort80 := NewReachability(k8s.AllPodStrings(), true)
                        // disallow all traffic to the X or Z namespace, BUT allow Y
                        reachabilityPort80.ExpectPeer(&Peer{Namespace: nsX, Pod:"a"}, &Peer{Namespace: nsX}, false)
                        reachabilityPort80.ExpectPeer(&Peer{Namespace: nsX, Pod:"a"}, &Peer{Namespace: nsZ}, false)
                        
                        //reachabilityPort80.ExpectPeer(&Peer{Namespace: nsZ}, &Peer{Namespace: nsX, Pod: "a"}, false)

                        ginkgo.By("Verify that port 80 is from ns y, but not from namespaces x and z")
                        // traffic to X over port 80 should succeed
                        ValidateOrFail(k8s, &TestCase{ToPort: 80, Protocol: v1.ProtocolTCP, Reachability: reachabilityPort80})

                        // traffic to X over port 81 should fail
                        ginkgo.By("Verify that port 81 is blocked for all namespaces including nxY")
                        reachabilityPort81 := NewReachability(k8s.AllPodStrings(), true)
                        reachabilityPort81.ExpectAllEgress(NewPodString(nsX, "a"), false)
                        ValidateOrFail(k8s, &TestCase{ToPort: 81, Protocol: v1.ProtocolTCP, Reachability: reachabilityPort81})

```

I think that expresses your policies intent.  

can someone you doublecheck that this isnt redundant to any other tests first b4 we merge ?

**Comment by yulng** [comment]
image
The diagram shows the intention of this test case

**Comment by yulng** [comment]
> 
According to your prompt, submitted again
I have analyzed the above several test cases related to ports, including ingress ports, ingress ports with ns restrictions, and egress ports, but without egress ports with ns restrictions

**Comment by yulng** [comment]
> @yulcan for a quick dev environment
> 
> 1. for infra locally w a cni .... Here's what I use https://github.com/jayunit100/k8sprototypes/tree/master/kind
> 
> It basically installs kind locally with a cni provider like calico or antrea . You can try both to confirm wether the errors you are seeing are bc of CNI issues or because of the code .......
> 
> 2. Now you'll also have to make your e2e.test binary.  You can do that by running the standard `make WHAT=test/e2e/e2e.test`  target in k8s.   Now you will see the e2e.test binary in your local directory...
> 3. finally you need to run that e2e.test binary against your cluster like -> %
> 
> then this...
> 
> ```
> test_binary=`find ./ -name e2e.test`
> 
> $test_binary --provider=local --kubeconfig=/home/ubuntu/.kube/config \
> --dump-logs-on-failure=false \
> --ginkgo.focus="egress access from namespace on one named port while blocking other namespaces and ports"
> ```
> 
> If you need help or want to pair there is a group of us ( me and @knabben) that meets every week to work on issues w new contributors we can send you an invite, email.me [jvyas@vmware.com](mailto:jvyas@vmware.com)

Could you run this test case in your environment? I installed the environment and reported "ImagePullBackOff"
image

**Comment by jayunit100** [comment]
It's probably worth debugging why that image won't pull down.  Maybe you need to docker save and then copy those images  into kind directly ? There is a way to load images into kind

**Comment by yulng** [comment]
> It's probably worth debugging why that image won't pull down. Maybe you need to docker save and then copy those images into kind directly ? There is a way to load images into kind
hello,[jayunit100](https://github.com/jayunit100)
My kind cluster is ready, and the calico node is also up, then run find./- Name e2e.test did not find the binary file for e2e.test

**Comment by yulng** [comment]
ping @jayunit100 
Could you review my question ? Thanks

**Comment by yulng** [comment]
@jayunit100

**Comment by jayunit100** [comment]
/reopen 

Sorry just saw

**Comment by jayunit100** [comment]
Ping me if this needs review again sorry I missed it.  Looks like bot closed it

**Comment by jayunit100** [comment]
Ping me if this needs review again sorry I missed it.  Looks like bot closed it

**Pull request closed without merging**